File: internal/cloud/portal/model_test.go
package portal

import (
	"encoding/json"
	"os"
	"os/exec"
	"path/filepath"
	"testing"

	"github.com/JoseAFlores777/ccp/internal/snapshot"
)

// El portal compara dos manifiestos en el navegador, así que su diff tiene que
// dar exactamente lo que da snapshot.Diff: si se desviaran, la línea de tiempo
// del portal diría una cosa y `ccp snapshot diff` otra sobre los mismos dos
// snapshots, y nadie sabría cuál mirar.
func TestDiffDelPortalIgualQueElDeGo(t *testing.T) {
	node, err := exec.LookPath("node")
	if err != nil {
		t.Skip("sin node: no se puede ejecutar el modelo del portal")
	}
	from := []snapshot.Item{
		{LPath: "ccp/ccp.yaml", Hash: "a1", Size: 100, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "ccp/profiles/work/overlay/CLAUDE.md", Hash: "b1", Size: 20, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "ccp/profiles/work/api_key", Hash: "c1", Size: 40, Mode: 0o600, Class: snapshot.ClassSecret},
		{LPath: "ccp/profiles/viejo/overlay/CLAUDE.md", Hash: "d1", Size: 10, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "claude/settings.json", Hash: "e1", Size: 30, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "claude/CLAUDE.md", Hash: "f1", Size: 30, Mode: 0o644, Class: snapshot.ClassAuthored},
	}
	to := []snapshot.Item{
		{LPath: "ccp/ccp.yaml", Hash: "a2", Size: 120, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "ccp/profiles/work/overlay/CLAUDE.md", Hash: "b1", Size: 20, Mode: 0o600, Class: snapshot.ClassAuthored},
		{LPath: "ccp/profiles/work/api_key", Hash: "c1", Size: 40, Mode: 0o600, Class: snapshot.ClassState},
		{LPath: "ccp/profiles/nuevo/overlay/settings.overlay.json", Hash: "g1", Size: 5, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "claude/settings.json", Hash: "e1", Size: 30, Mode: 0o644, Class: snapshot.ClassAuthored},
		{LPath: "project/github.com~acme~web/.claude/settings.json", Hash: "h1", Size: 8, Mode: 0o644, Class: snapshot.ClassAuthored},
	}
	payload := map[string]any{
		"from":     map[string]any{"items": from},
		"to":       map[string]any{"items": to},
		"expected": snapshot.Diff(from, to),
		"profiles": []string{"nuevo", "viejo", "work"},
	}
	b, err := json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(t.TempDir(), "modelo.json")
	if err := os.WriteFile(path, b, 0o600); err != nil {
		t.Fatal(err)
	}
	out, err := exec.Command(node, "model_test.mjs", path).CombinedOutput()
	t.Logf("%s", out)
	if err != nil {
		t.Fatalf("el modelo del portal no coincide con snapshot.Diff: %v", err)
	}
}

File: internal/cloud/portal/model_test.mjs
// Comprueba que el modelo del portal (diff, perfiles, áreas) dice lo mismo que
// internal/snapshot. Los vectores los genera model_test.go.
import { readFileSync } from 'node:fs';
import { diff, profilesOf, areaOf, resumen } from './web/js/model.js';

const v = JSON.parse(readFileSync(process.argv[2], 'utf8'));
let fallos = 0;
const check = (nombre, ok, extra) => {
  if (!ok) { fallos++; console.error('FALLA', nombre, extra ?? ''); } else { console.log('ok  ', nombre); }
};

const norm = (cs) => cs.map((c) => [
  c.lpath, c.kind,
  c.from ? `${c.from.hash}/${c.from.mode}/${c.from.class}` : '-',
  c.to ? `${c.to.hash}/${c.to.mode}/${c.to.class}` : '-',
].join('|'));

const got = norm(diff(v.from.items, v.to.items));
const want = norm(v.expected);
check('diff igual que snapshot.Diff', JSON.stringify(got) === JSON.stringify(want), '\n  js: ' + got.join('\n  js: ') + '\n  go: ' + want.join('\n  go: '));

const perfiles = [...new Set([...profilesOf(v.from), ...profilesOf(v.to)])].sort();
check('perfiles del manifiesto', JSON.stringify(perfiles) === JSON.stringify(v.profiles), perfiles);

check('área de un perfil', areaOf('ccp/profiles/work/overlay/CLAUDE.md') === 'perfil work');
check('área global', areaOf('claude/settings.json') === 'global');
check('área de ccp', areaOf('ccp/ccp.yaml') === 'ccp');
check('área de un proyecto', areaOf('projects/github.com~acme~web/.claude/settings.json') === 'proyecto github.com~acme~web');

const r = resumen(diff(v.from.items, v.to.items));
check('resumen cuenta por tipo', r.added === 2 && r.removed === 2 && r.modified === 3, JSON.stringify(r));

process.exit(fallos === 0 ? 0 : 1);
